File: Cpp/958_Check_Completeness_of_a_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCompleteTree(TreeNode* root) {
        int count = 0;

        return recursive(root, 1, &count) == count;
    }
    
    unsigned long long recursive(TreeNode* node, unsigned long long index, int* count) {
        if (!node) return 0;
        
        ++(*count);
        return max({index, recursive(node->left, 2 * index, count), recursive(node->right, 2 * index + 1, count)});
    }
};

File: Kotlin/958_Check_Completeness_of_a_Binary_Tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    var count = 0
    fun isCompleteTree(root: TreeNode?): Boolean {
        return recursive(root, 1) == count
    }

    fun recursive(node: TreeNode?, index: Int): Int {
        if (node == null) return 0
        
        ++count
        return maxOf(maxOf(index, recursive(node.left, 2 * index)), recursive(node.right, 2 * index + 1))
    }
}